File: src/vm/stack.ts
// others
import type { Value } from "../bytecode";
import type { VMState } from "./state";

// ----------------------------------------------------------------------

export function push(state: VMState, val: Value) {
	state.stack.push(val);
}

export function pop(state: VMState): Value {
	if (state.stack.length === 0) throw new Error("Stack underflow");
	return state.stack.pop() as Value;
}

export function peek(state: VMState, dist: number): Value {
	return state.stack[state.stack.length - 1 - dist] as Value;
}

File: zig/src/vm/stack.zig
const state_mod = @import("state.zig");
const Value = state_mod.Value;
const VMState = state_mod.VMState;

pub fn push(vm: *VMState, v: Value) !void {
    try vm.stack.append(vm.allocator, v);
}

pub fn pop(vm: *VMState) Value {
    return vm.stack.pop().?;
}

pub fn peek(vm: *const VMState, distance: usize) Value {
    const len = vm.stack.items.len;
    return vm.stack.items[len - 1 - distance];
}
